> Started Windows CI on this PR - https://ci-hadoop.apache.org/view/Hadoop/job/hadoop-qbt-trunk-java8-win10-x86_64/654/

@kevincai the Windows CI failed to build the HDFS native client due to the following error -
```
     [exec] Link:
     [exec]   C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64\link.exe /ERRORREPORT:QUEUE /OUT:"C:\hadoop\hadoop-hdfs-project\hadoop-hdfs-native-client\target\native\main\native\libhdfspp\tests\RelWithDebInfo\get_jni_test.exe" /INCREMENTAL:NO /NOLOGO ..\..\..\..\lib\RelWithDebInfo\gmock_main.lib ..\..\..\..\bin\RelWithDebInfo\hdfs.lib "C:\Java\zulu8.62.0.19-ca-jdk8.0.332-win_x64\lib\jvm.lib" kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /DEBUG /PDB:"C:/hadoop/hadoop-hdfs-project/hadoop-hdfs-native-client/target/native/main/native/libhdfspp/tests/RelWithDebInfo/get_jni_test.pdb" /SUBSYSTEM:CONSOLE /TLBID:1 /DYNAMICBASE /NXCOMPAT /IMPLIB:"C:/hadoop/hadoop-hdfs-project/hadoop-hdfs-native-client/target/native/main/native/libhdfspp/tests/RelWithDebInfo/get_jni_test.lib" /MACHINE:X64  /machine:x64 get_jni_test.dir\RelWithDebInfo\libhdfs_getjni_test.obj
     [exec] jvm.lib(jvm.dll) : error LNK2005: JNI_CreateJavaVM already defined in libhdfs_getjni_test.obj [C:\hadoop\hadoop-hdfs-project\hadoop-hdfs-native-client\target\native\main\native\libhdfspp\tests\get_jni_test.vcxproj]
     [exec]      Creating library C:/hadoop/hadoop-hdfs-project/hadoop-hdfs-native-client/target/native/main/native/libhdfspp/tests/RelWithDebInfo/get_jni_test.lib and object C:/hadoop/hadoop-hdfs-project/hadoop-hdfs-native-client/target/native/main/native/libhdfspp/tests/RelWithDebInfo/get_jni_test.exp
     [exec] C:\hadoop\hadoop-hdfs-project\hadoop-hdfs-native-client\target\native\main\native\libhdfspp\tests\RelWithDebInfo\get_jni_test.exe : fatal error LNK1169: one or more multiply defined symbols found [C:\hadoop\hadoop-hdfs-project\hadoop-hdfs-native-client\target\native\main\native\libhdfspp\tests\get_jni_test.vcxproj]
```

The error seems valid. You've defined `JNI_CreateJavaVM` in [libhdfs_getjni_test.cc](https://github.com/apache/hadoop/pull/6969/files#diff-589ac64cc250ab38f319997b6247594d71d19bdb4b73ddea59730cfab39611be). This method is already present in the `jvm.dll`.

Were you trying to override the definition for `JNI_CreateJavaVM`? If so, you shouldn't be linking `jvm.dll` to `get_jni_test` in `CMakeLists.txt`.

--- a/hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfs/CMakeLists.txt
+++ b/hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfs/CMakeLists.txt
@@ -46,6 +46,7 @@ set(HDFS_SOURCES
 # link to hdfs's publicly linked libraries
 # (like jvm)
 add_library(hdfs_obj OBJECT ${HDFS_SOURCES})
+set_target_properties(hdfs_obj PROPERTIES POSITION_INDEPENDENT_CODE ON)
 hadoop_add_dual_library(hdfs
     $<TARGET_OBJECTS:hdfs_obj>
     $<TARGET_OBJECTS:x_platform_obj>

--- a/hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfs/CMakeLists.txt
+++ b/hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfs/CMakeLists.txt
@@ -40,6 +40,11 @@ set(HDFS_SOURCES
     ${OS_DIR}/mutexes.c
     ${OS_DIR}/thread_local_storage.c
 )
+# We want to create an object library for hdfs
+# so that we can reuse it for the targets
+# (like get_jni_test), where we don't wish to
+# link to hdfs's publicly linked libraries
+# (like jvm)
 add_library(hdfs_obj OBJECT ${HDFS_SOURCES})
 hadoop_add_dual_library(hdfs
     $<TARGET_OBJECTS:hdfs_obj>

--- a/hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfs/CMakeLists.txt
+++ b/hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfs/CMakeLists.txt
@@ -32,13 +32,17 @@ include_directories(
     ../libhdfspp/lib
 )
 
-hadoop_add_dual_library(hdfs
+set(HDFS_SOURCES
     exception.c
     jni_helper.c
     hdfs.c
     jclasses.c
     ${OS_DIR}/mutexes.c
     ${OS_DIR}/thread_local_storage.c
+)
+add_library(hdfs_obj OBJECT ${HDFS_SOURCES})
+hadoop_add_dual_library(hdfs
+    $<TARGET_OBJECTS:hdfs_obj>
     $<TARGET_OBJECTS:x_platform_obj>
     $<TARGET_OBJECTS:x_platform_obj_c_api>
 )

--- a/hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfspp/tests/CMakeLists.txt
+++ b/hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfspp/tests/CMakeLists.txt
@@ -74,8 +74,17 @@ add_executable(uri_test uri_test.cc)
 target_link_libraries(uri_test common gmock_main ${CMAKE_THREAD_LIBS_INIT})
 add_memcheck_test(uri uri_test)
 
+get_target_property(HDFS_STATIC_LIBS_NO_JVM hdfs_static LINK_LIBRARIES)
+list(REMOVE_ITEM HDFS_STATIC_LIBS_NO_JVM ${JAVA_JVM_LIBRARY})
+
 add_executable(get_jni_test libhdfs_getjni_test.cc)
-target_link_libraries(get_jni_test gmock_main hdfs_static ${CMAKE_THREAD_LIBS_INIT})
+target_link_libraries(get_jni_test
+    gmock_main
+    $<TARGET_OBJECTS:hdfs_obj>
+    $<TARGET_OBJECTS:x_platform_obj>
+    $<TARGET_OBJECTS:x_platform_obj_c_api>
+    ${HDFS_STATIC_LIBS_NO_JVM}
+    ${CMAKE_THREAD_LIBS_INIT})
 add_memcheck_test(get_jni get_jni_test)
 
 add_executable(remote_block_reader_test remote_block_reader_test.cc)

--- a/hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfspp/tests/CMakeLists.txt
+++ b/hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfspp/tests/CMakeLists.txt
@@ -74,9 +74,11 @@ add_executable(uri_test uri_test.cc)
 target_link_libraries(uri_test common gmock_main ${CMAKE_THREAD_LIBS_INIT})
 add_memcheck_test(uri uri_test)
 
+# We want to link to all the libraries of hdfs_static library,
+# except jvm.lib since we want to override some of the functions
+# provided by jvm.lib.
 get_target_property(HDFS_STATIC_LIBS_NO_JVM hdfs_static LINK_LIBRARIES)
 list(REMOVE_ITEM HDFS_STATIC_LIBS_NO_JVM ${JAVA_JVM_LIBRARY})
-
 add_executable(get_jni_test libhdfs_getjni_test.cc)
 target_link_libraries(get_jni_test
     gmock_main

--- a/hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfspp/tests/libhdfs_getjni_test.cc
+++ b/hadoop-hdfs-project/hadoop-hdfs-native-client/src/main/native/libhdfspp/tests/libhdfs_getjni_test.cc
@@ -38,6 +38,11 @@ DECLSPEC jint JNICALL JNI_CreateJavaVM(JavaVM**, void**, void*) {
     return 1;
 }
 
+// hook the jvm runtime function. expect always failure
+DECLSPEC jint JNICALL JNI_GetCreatedJavaVMs(JavaVM**, jsize, jsize*) {
+    return 1;
+}
+
 TEST(GetJNITest, TestRepeatedGetJNIFailsButNoCrash) {
     // connect to nothing, should fail but not crash
     EXPECT_EQ(NULL, hdfsConnectNewInstance(NULL, 0));

